README.md
:warning: **THIS IS EXPERIMENTAL SOFTWARE** :warning:

# ![normalmap.js](doc/gfx/logo-rendered.png)
[API](doc/API.md) |
[Example](#example) |
[Installation](#installation) |
[Changelog](doc/CHANGELOG.md) |
[License](#license) (MIT) |
[Questions?](#questions)

normalmap.js is a library for creating simple interactive lighting effects.

- Built on top of WebGL
- Simple to use
- Small (~7kb gzipped)
- Fast enough to run on modern smartphones
- There is a tiny jQuery plugin for simple usage
- Covered by [automated end-to-end tests](https://29a.ch/sandbox/2016/normalmap.js/tests/)

# Demos
[![Golden Dragon](doc/gfx/dragon.gif)](https://29a.ch/sandbox/2016/normalmap.js/)

Did you find a cool use for normalmap.js? [Let me know!](https://29a.ch/about)

# Example
```javascript
var lights = normalMap({
    canvas: canvas,
    normalMap: normalMap,
    baseColor: [0.5, 0.4, 0.5]
});

lights.clear();

var position = normalMap.vec3(0.5, 0.5, 1.0); // dead center
var color = normalMap.vec3(1.0, 1.0, 1.0); // bright white
lights.addPointLight(position, color);
```

# Installation
With NPM:
```
$ npm install normalmap
```

```javascript
var normalmap = require('normalmap');
```

With Bower:
```$ bower install normalmap```

Or just download the latest zip file form the [releases page](https://github.com/jwagner/normalmap.js/releases).

Note that normalmap.js won't work with file:// urls. You will need to serve
images from a webserver. See Richard Daveys [phaser introduction](http://phaser.io/tutorials/getting-started) for an explanations for this.

# Browser Support

Normalmap.js should work in all browsers that [supports WebGL](http://caniuse.com/#feat=webgl).

# API Documentation

[doc/API.md](doc/API.md)

# Changelog
[doc/CHANGELOG.md](doc/CHANGELOG.md)

# Creating normal maps
[NormalMap-Online](http://cpetry.github.io/NormalMap-Online/)

# Questions?
Create a github issue or tweet to @29a_ch and please be a bit patient. :)

# License
Copyright © 2016 Jonas Wagner.

The **source code** of this project licensed under the [MIT License](LICENSE) (enclosed).

The images and textures used for demonstration purposes are **not** licensed under the MIT License and belong to their respective owners.


doc/CHANGELOG.md
# ![normalmap.js](gfx/logo-rendered.png)

# Changelog

## master

## 0.0.2 (2016-03-13)

* Cleaned up console.logs
* Fixed sRGB FBO initialization

## 0.0.1 (2016-03-13)

* Release with newer version of npm to prevent browserify transform mangling
* Move debounce to devDependencies


## 0.0.0 (2016-03-13)

Initial experimental public release.


tests/test.js
/* globals $, normalmap */
(function(){

var vec3 = normalmap.vec3;
var gold = vec3(1.000, 0.766, 0.336);

var lastTest = $.when(true);
var nTests = 0;
var nTestsSucceeded = 0;

function test(name, body){
    // if(name !== 'material metalness roughness') return;
    var canvas = $('<canvas>')[0];
    var expectedSrc = 'expected/' + name.replace(/ /g, '-') + '.png';
    var expected;
    // stupid jquery defereds
    return lastTest = lastTest.then(function(){}, function(){ return $.when(true);})
    .then(function(){
        nTests++;
        return $.fn.normalmap._loadImage(expectedSrc);
    })
    .then(function(expected_){
        expected = expected_;
    }, function(e){ return $.when(e); })
    .then(function(){
        return body(canvas);
    }).then(function(){
        var difference = diff(canvas, expected);
        var failed = difference.total > 0.005;
        if(!failed) nTestsSucceeded++;
        $('body')
            .append(
                $('<h2>')
                    .text(name + ' (' + Math.floor(difference.total * 100000)/100000 + ')')
                    .css('color', failed ? 'red' : 'green')
            )
            .append(
                $('<div>')
                    // copy the canvas to work around the webgl context limit
                    .append(copyCanvas(canvas))
                    .append(expected ? expected : 'missing ' + expectedSrc)
                    .append(difference.canvas)
            );
    });
}

function copyCanvas(a){
    var c = document.createElement('canvas');
    var ctx = c.getContext('2d');
    c.width = a.width;
    c.height = a.height;
    ctx.drawImage(a, 0, 0);
    return c;
}

// currently ignores alpha
function diff(a, b){
    var c = document.createElement('canvas');
    c.width = a.width;
    c.height = a.height;
    var ctx = c.getContext('2d');
    ctx.clearRect(0, 0, c.width, c.height);
    ctx.drawImage(a, 0, 0);
    if(!b || !b.complete) return {
        canvas: c,
        total: 1
    };
    var aData = ctx.getImageData(0, 0, c.width, c.height).data;
    ctx.clearRect(0, 0, c.width, c.height);
    ctx.drawImage(b, 0, 0);
    var bImgData = ctx.getImageData(0, 0, c.width, c.height);
    var bData = bImgData.data;
    var total = 0;
    for(var i = 0; i < bData.length; i+=4){
        for(var j = 0; j < 4; j++) {
            var aAlpha = j < 3 ? aData[i+3]/256 : 1;
            var bAlpha = j < 3 ? bData[i+3]/256 : 1;
            var d = (aData[i+j]/256*aAlpha-bData[i+j]/256*bAlpha);
            total += d*d;
        }
        bData[i] = 127 + aData[i]*0.5 - bData[i]*0.5;
        bData[i+1] = 127 + aData[i+1]*0.5 - bData[i+1]*0.5;
        bData[i+2] = 127 + aData[i+2]*0.5 - bData[i+2]*0.5;
    }
    ctx.putImageData(bImgData, 0, 0);
    return {
        canvas: c,
        total: total/(c.width*c.height)
    };
}

test('simple normals', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal.png'
    }).then(function(lights){
        lights.addPointLight(
            vec3(0.5, 0.5, 0.5),
            vec3(1.0, 1.0, 1.0)
        );
        lights.render();
    });
});

test('many lights', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal.png',
    }).then(function(lights){
        for(var i = 0; i < 10; i++){
            lights.addPointLight(
                vec3(0.5, 0.5, 0.5),
                vec3(0.1, 0.1, 0.1)
            );
        }
        lights.render();
    });
});

test('simple normals top left', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal.png'
    }).then(function(lights){
        lights.addPointLight(
            vec3(0.0, 0.0, 0.5),
            vec3(1.0, 1.0, 1.0)
        );
        lights.render();
    });
});

test('simple normals red', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal.png'
    }).then(function(lights){
        lights.addPointLight(
            vec3(0.5, 0.5, 0.5),
            vec3(1.0, 0, 0)
        );
        lights.render();
    });
});

test('simple normals directional', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal.png'
    }).then(function(lights){
        lights.addDirectionalLight(
            vec3(0.5, 0.5, -0.1),
            vec3(1, 1, 1)
        );
        lights.render();
    });
});

test('four lights', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal.png'
    }).then(function(lights){
        lights.addPointLight(
            vec3(-1.0, 0.4, 0.1),
            vec3(4, 1.2, 0.1)
        );
        lights.addPointLight(
            vec3(2.0, 0.4, 0.1),
            vec3(0.1, 1, 4)
        );
        lights.addPointLight(
            vec3(0.4, -1.0, 0.1),
            vec3(4, 0.1, 1.2)
        );
        lights.addPointLight(
            vec3(0.4, 2, 0.1),
            vec3(0.1, 1, 4)
        );
        lights.render();
    });
});

test('metal', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal.png',
        metalness: 1.0,
        baseColor: gold,
        roughness: 0.1
    }).then(function(lights){
        lights.addPointLight(
            vec3(0.5, 0.5, 1),
            vec3(1.0, 1.0, 1.0)
        );
        lights.render();
    });
});

test('plastic', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal.png',
        metalness: 0.0,
        baseColor: gold,
        roughness: 0.1
    }).then(function(lights){
        lights.addPointLight(
            vec3(0.5, 0.5, 1),
            vec3(1.0, 1.0, 1.0)
        );
        lights.render();
    });
});

test('roughness', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal.png',
        metalness: 0.0,
        baseColor: gold,
        roughness: 0.5
    }).then(function(lights){
        lights.addPointLight(
            vec3(0.5, 0.5, 1),
            vec3(1.0, 1.0, 1.0)
        );
        lights.render();
    });
});

test('baseColorMap', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal.png',
        metalness: 0.0,
        baseColor: gold,
        baseColorMap: 'checker.png',
        roughness: 0.5
    }).then(function(lights){
        lights.addPointLight(
            vec3(0.5, 0.5, 1),
            vec3(1.0, 1.0, 1.0)
        );
        lights.render();
    });
});

test('ambientMap', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal.png',
        metalness: 0.0,
        baseColor: gold,
        ambientMap: 'checker-r.png',
        roughness: 0.5
    }).then(function(lights){
        lights.addPointLight(
            vec3(0.5, 0.5, 1),
            vec3(1.0, 1.0, 1.0)
        );
        lights.render();
    });
});

test('ambientMapAlpha', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal.png',
        metalness: 0.0,
        baseColor: gold,
        ambientMap: 'checker-a.png',
        roughness: 0.5
    }).then(function(lights){
        lights.addPointLight(
            vec3(0.5, 0.5, 1),
            vec3(1.0, 1.0, 1.0)
        );
        lights.render();
    });
});

test('ambientMapAlpha singlePass', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal.png',
        metalness: 0.0,
        baseColor: gold,
        ambientMap: 'checker-a.png',
        roughness: 0.5,
        singlePass: true
    }).then(function(lights){
        lights.addPointLight(
            vec3(0.5, 0.5, 1),
            vec3(1.0, 1.0, 1.0)
        );
        lights.addPointLight(
            vec3(0.5, 0.5, 1),
            vec3(1.0, 1.0, 1.0)
        );
    });
});

test('ambientMapAlpha many lights', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal.png',
        metalness: 0.0,
        baseColor: gold,
        ambientMap: 'checker-a.png',
        roughness: 0.5
    }).then(function(lights){
        for(var i = 0; i < 10; i++){
            lights.addPointLight(
                vec3(0.5, 0.5, 1),
                vec3(0.1, 0.1, 0.1)
            );
        }
        lights.render();
    });
});

test('material metalness roughness', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal-sharp.png',
        baseColor: gold,
        materialMap: 'material.png'
    }).then(function(lights){
        lights.addPointLight(
            vec3(0.5, 0.5, 1),
            vec3(1.0, 1.0, 1.0)
        );
        lights.render();
    });
});

test('material occlusion', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal.png',
        baseColor: gold,
        materialMap: 'checker-b.png'
    }).then(function(lights){
        lights.addPointLight(
            vec3(0.5, 0.5, 1),
            vec3(1.0, 1.0, 1.0)
        );
        lights.render();
    });
});

test('repeat', function(canvas){
    canvas.width = 512;
    canvas.height = 512;
    return $(canvas).normalmap({
        normalMap: 'normal.png',
        repeat: true
    }).then(function(lights){
        lights.addPointLight(
            vec3(0.5, 0.5, 0.5),
            vec3(1.0, 1.0, 1.0)
        );
        lights.render();
    });
});

test('subSurfaceScattering 1', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal-sharp.png',
        baseColor: gold,
        subSurfaceScattering: 0,
        metalness: 0,
        roughness: 0.0
    }).then(function(lights){
        lights.addPointLight(
            vec3(0.5, 0.25, 1),
            vec3(1.0, 1.0, 1.0)
        );
        lights.render();
    });
});

test('subSurfaceScattering 3', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal-sharp.png',
        baseColor: gold,
        subSurfaceScattering: 3,
        metalness: 0,
        roughness: 0.5
    }).then(function(lights){
        lights.addPointLight(
            vec3(0.5, 0.25, 1),
            vec3(1.0, 1.0, 1.0)
        );
        lights.render();
    });
});


test('npot', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal-npot.png'
    }).then(function(lights){
        lights.addPointLight(
            vec3(0.5, 0.5, 0.5),
            vec3(1.0, 1.0, 1.0)
        );
        lights.render();
    });
});

test('configure values', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal.png',
    }).then(function(lights){
        lights.configure({
            baseColor: gold,
            metalness: 0,
            roughness: 0.5
        });
        lights.addPointLight(
            vec3(0.5, 0.5, 0.5),
            vec3(1.0, 1.0, 1.0)
        );
        lights.render();
    });
});

test('configure normalMap', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal.png',
    }).then(function(lights){
        return $.fn.normalmap._loadImage('normal-sharp.png').then(function(n){
            lights.configure({
                normalMap: n
            });
            lights.addPointLight(
                vec3(0.5, 0.5, 0.5),
                vec3(1.0, 1.0, 1.0)
            );
            lights.render();
        });
    });
});

test('configure normalMap', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal.png',
    }).then(function(lights){
        return $.fn.normalmap._loadImage('normal-sharp.png').then(function(n){
            lights.configure({
                normalMap: n
            });
            lights.addPointLight(
                vec3(0.5, 0.5, 0.5),
                vec3(1.0, 1.0, 1.0)
            );
            lights.render();
        });
    });
});

test('configure materialMap', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal-sharp.png',
        baseColor: vec3(0.5, 0.5, 0.5)
    }).then(function(lights){
        return $.fn.normalmap._loadImage('material.png').then(function(img){
            lights.configure({
                materialMap: img
            });
            lights.addPointLight(
                vec3(0.5, 0.5, 1.0),
                vec3(1.0, 1.0, 1.0)
            );
            lights.render();
        });
    });
});

test('configure ambientMap', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal-sharp.png',
        baseColor: vec3(0.5, 0.5, 0.5)
    }).then(function(lights){
        return $.fn.normalmap._loadImage('checker.png').then(function(img){
            lights.configure({
                ambientMap: img
            });
            lights.addPointLight(
                vec3(0.5, 0.5, 1.0),
                vec3(1.0, 1.0, 1.0)
            );
            lights.render();
        });
    });
});

test('antiAliasing', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal-sharp.png',
        baseColor: vec3(1.0, 1.0, 1.0),
        roughness: 0.05,
        antiAliasing: true
    }).then(function(lights){
        lights.addPointLight(
            vec3(0.5, 0.0, 1.0),
            vec3(1.0, 1.0, 1.0)
        );
        lights.render();
    });
});

test('muliple render', function(canvas){
    return $(canvas).normalmap({
        normalMap: 'normal.png',
        baseColor: vec3(0.5, 0.5, 0.5)
    }).then(function(lights){
        lights.addPointLight(
            vec3(0.5, 0.5, 1.0),
            vec3(1.0, 0.0, 0.0)
        );
        lights.render();
        lights.addPointLight(
            vec3(0.5, 0.5, 1.0),
            vec3(0.0, 1.0, 0.0)
        );
        lights.render();
    });
});

lastTest.then(function(){
    var success = nTestsSucceeded == nTests;
    $('body').prepend(
        $('<div class=test-results>')
            .text('summary: ' + nTestsSucceeded + '/' + nTests + ' succeeded')
            .css('color', success ? 'green' : 'red')
            .addClass('test-results-' + (success ? 'passed' : 'failed'))
    );
});

})();


scripts/run-tests.js
var webdriverio = require('webdriverio');

var browsers = [
    {
        browserName: 'MicrosoftEdge',
        'platform': 'Windows 10',
        'version': '20.10240',
    },
    {
        browserName: 'internet explorer',
        'platform': 'Windows 10',
        'version': '11.0',
    },
    {
        browserName: 'iphone',
        platform: 'OS X 10.10',
        version: '9.2',
        deviceName: 'iPhone 6 Plus'
    }
];


function runTests(browser){
    var options = {
        host: 'ondemand.saucelabs.com',
        port: 80,
        user: process.env.SAUCE_USER,
        key: process.env.SAUCE_KEY,
        desiredCapabilities: browser,
        // logLevel: 'verbose'
    };
    var name = browser.browserName + ' ' + browser.version;
    var client = webdriverio.remote(options);
    console.log('starting ' + name);
    var timeout = 180*1000;
    client
        .init()
        .url('https://29a.ch/sandbox/2016/normalmap.js/tests/')
        .waitForVisible('.test-results', timeout)
        .element('.test-results-passed')
        .then(function(el){
            console.log(name + ' passed');
        }, function(e){
            console.log(name + ' passed', e);
        })
        .end();
}

browsers.forEach(runTests);


scripts/run-tests.sh
#!/bin/bash
cd $(dirname $0)
source ~/.config/saucelabs.sh
echo $SAUCE_USER
node ./run-tests.js


src/fbo.js
function FBO(gl, width, height, type, format, depth){
    width = width || gl.drawingBufferWidth;
    height = height || gl.drawingBufferHeight;
    this.width = width;
    this.height = height;
    this.gl = gl;

    this.framebuffer = gl.createFramebuffer();
    gl.bindFramebuffer(gl.FRAMEBUFFER, this.framebuffer);

    this.texture = gl.createTexture();
    gl.bindTexture(gl.TEXTURE_2D, this.texture);
    gl.texImage2D(gl.TEXTURE_2D, 0, format || gl.RGBA, width, height, 0, format || gl.RGBA, type || gl.UNSIGNED_BYTE, null);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);

    if(depth){
        this.depth = gl.createRenderbuffer();
        gl.bindRenderbuffer(gl.RENDERBUFFER, this.depth);
        gl.renderbufferStorage(gl.RENDERBUFFER, gl.DEPTH_COMPONENT16, width, height);
        gl.framebufferRenderbuffer(gl.FRAMEBUFFER, gl.DEPTH_ATTACHMENT, gl.RENDERBUFFER, this.depth);
    }
    gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, this.texture, 0);

    this.supported = gl.checkFramebufferStatus(gl.FRAMEBUFFER) === gl.FRAMEBUFFER_COMPLETE;

    gl.viewport(0, 0, width, height);
    gl.bindTexture(gl.TEXTURE_2D, null);
    gl.bindRenderbuffer(gl.RENDERBUFFER, null);
    gl.bindFramebuffer(gl.FRAMEBUFFER, null);
}
FBO.prototype = {
    bind: function () {
        this.gl.bindFramebuffer(this.gl.FRAMEBUFFER, this.framebuffer);
    },
    unbind: function() {
        this.gl.bindFramebuffer(this.gl.FRAMEBUFFER, null);
    }
};
module.exports = FBO;


gulpfile.js
var gulp = require('gulp');
var browserSync = require('browser-sync');
var browserify = require('browserify');
var source = require('vinyl-source-stream');
var buffer = require('vinyl-buffer');
var glob = require('glob');
var uglify = require('gulp-uglify');
var rename = require('gulp-rename');
var gutil = require('gulp-util');
var newer = require('gulp-newer');
var data = require('gulp-data');
var imageminOptipng = require('imagemin-optipng');
var jade = require('gulp-jade');
var yazl = require('yazl');
var imageminPngquant = require('imagemin-pngquant');
var rsync = require('rsyncwrapper');
var debounce = require('debounce');
var asyncDone = require('async-done');

var meta = require('./package.json');

const BUILD_DIR_DEV = 'build/dev/';
const BUILD_DIR_RELEASE = 'build/release/';


function serve(openBrowser){
    browserSync({
      server: {
        baseDir: 'build/dev/demos/',
      },
      ghostMode: false,
      open: openBrowser,
    });
    // fugly hack to work around reloads being triggered before
    // writes are complete
    var reload = debounce(browserSync.reload, 1000);
    gulp.watch(['src/*.js', 'src/shaders/*.glsl'], ['build']);
    gulp.watch(['demos/*.jade'], ['buildHtml']);
    gulp.watch(['demos/gfx/*/*.png', 'demos/*.{js,css}'], ['copyAssets']);
    gulp.watch(['**/*'], {cwd: 'build/dev/demos/'}, reload);
    gulp.watch(['**/*'], {cwd: 'dist/tests/'}, reload);
}

function buildJavascript(release, dest) {
    console.log('building ' + meta.name + '@' + meta.version);
    var b = browserify({
            entries: meta.main,
            standalone: meta.name,
            debug: true,
            // insertGlobals: !release
        });

    var stream = b.bundle()
        .on('error', function(error){
            console.log(error.toString());
            stream.end();
        })
        .pipe(source(meta.name + '.js'))
        .pipe(buffer())
        // .pipe(sourcemaps.init({loadMaps: true}))
        .on('error', gutil.log)
        .pipe(gulp.dest(dest));

    if(release) {
        stream = stream
            .pipe(uglify({
                compress: {
                    unsafe: true,
                    screw_ie8: true
                },
                mangle: {
                    screw_ie8: true
                },
                output: {
                    screw_ie8: true
                }
            }))
            .pipe(rename({extname: '.min.js'}))
            .pipe(gulp.dest(dest));
    }
    return stream;
}
gulp.task('buildJavascript', buildJavascript.bind(null, false, BUILD_DIR_DEV));

function copyImages(src, dest, compress){
    var stream = gulp.src(src, {base: '.'})
            .pipe(newer(dest));

    if(compress === 'quantize'){
        stream = stream.pipe(imageminPngquant()());
    }
    else if(compress === 'optimize') {
        stream = stream.pipe(imageminOptipng({optimizationLevel: 7})());
    }

    return stream.pipe(gulp.dest(dest));
}

function copyAssets(release, dest){
    console.log('copyAssets');
    return streamsToPromise([
        copyImages(['demos/gfx/*/{ambient,baseColor,material}.png'], dest, release && 'quantize'),
        copyImages(['demos/gfx/*/normal.png'], dest, release && 'optimize'),
        gulp.src(['demos/*.{js,css,gif}'], {base: '.'}).pipe(gulp.dest(dest)),
        gulp.src(['src/jquery.normalmap.js']).pipe(gulp.dest(dest)),
    ]).then(function(){
        return streamToPromise(
            gulp.src([dest + '?(jquery.)normalmap.js']).pipe(gulp.dest(dest + 'demos/'))
        );
    });
}
gulp.task('copyAssets', copyAssets.bind(this, false, BUILD_DIR_DEV));


function buildHtml(release, dest){
    console.log('building html', release);
    var stream = gulp.src(['demos/*.jade', '!demos/_*.jade'])
        .pipe(data(function(file){
            return {
                meta: meta,
                release: release,
                baseName: gutil.replaceExtension(path.basename(file.path), '.html').replace('index.html', ''),
                fileRev: release ? fileRev.bind(this, dest) : identity,
            };
        }))
        .pipe(jade({
            pretty: true
        }))
        .on('error', function(e){
            gutil.log(e);
            stream.end();
        })
        .pipe(gulp.dest(dest));
    return stream;
}
gulp.task('buildHtml', ['copyAssets'], buildHtml.bind(this, false, BUILD_DIR_DEV + 'demos'));

function build(release, dest, releaseJavascript){
    return streamToPromise(buildJavascript(release || releaseJavascript, dest))
        .then(copyAssets.bind(null, release, dest))
        .then(buildHtml.bind(null, release, dest + 'demos/'));
}
gulp.task('build', build.bind(this, false, BUILD_DIR_DEV));
gulp.task('buildRelease', build.bind(this, true, BUILD_DIR_RELEASE));

gulp.task('open', ['build'], function(){
    return serve(true);
});

gulp.task('default', ['build'], function() {
    return serve();
});

gulp.task('release', ['buildRelease'], function(done){
    rsync({
        src: 'build/release/demos/',
        recursive: true,
        exclude: ['.*'],
        dest: '/var/www/static/sandbox/2016/normalmap.js/',
        host: '29a.ch',
        port: '22',
        args: ['--copy-links'],
        dryRun: false
    }, function(error, stdout, stderr, cmd){
        console.log(cmd);
        if(error) {
            console.trace(error);
            console.warn(stderr);
        }
        console.log(stdout);
        done();
    });
});

gulp.task('dist', ['buildRelease'], function(){
    return gulp.src(['build/release/*.js'])
        .pipe(gulp.dest('dist/'));
});

gulp.task('zip', function(){
    // Symlink handling in gulp/vinyl-fs is not helpful
    // so I'm using yazl directly
    return build(false, BUILD_DIR_DEV, true)
        .then(function(){
            var name = meta.name + '-' + meta.version;
            var zipFile = new yazl.ZipFile();
            var globOptions = {
                dot: false,
                follow: true,
                mark: true,
                ignore: ['**/*.cache-*']
            };

            glob(BUILD_DIR_DEV + '**/*', globOptions, function(e, files){
                if(e){
                    stream.emit('error', e);
                }
                files.forEach(function(file){
                    if(file[file.length-1] == '/') return;
                    console.log(file);
                    zipFile.addFile(file, file.replace(BUILD_DIR_DEV, meta.name + '/'));
                });
                zipFile.end();
            });

            ['README.md', 'doc/API.md', 'doc/CHANGELOG.md'].forEach(function(file){
               zipFile.addFile(file, meta.name + '/' + file.replace('doc/', ''));
            });

            var stream = zipFile.outputStream
                .pipe(fs.createWriteStream('build/zip/' + name + '.zip'));

            return streamToPromise(stream);
    });
});

function streamToPromise(stream){
    return new Promise(function(resolve, reject){
        asyncDone(identity.bind(null, stream), function(e){
            if(e) return reject(e);
            resolve(stream);
        });
    });
}

function streamsToPromise(streams){
    return Promise.all(streams.map(streamToPromise));
}

var fs = require('fs'),
    path = require('path'),
    crypto = require('crypto');

function fileRev(root, fileName){
    var filePath = path.join(root, fileName);
    if(!fs.existsSync(filePath)){
        return fileName;
    }
    var buffer = fs.readFileSync(filePath),
        hash = crypto.createHash('md5');

    hash.update(buffer);

    var digest = hash.digest('hex').slice(0, 16);
    var fileNameRev = fileName.replace(/([^.\/]+)$/, 'cache-' + digest + '.$1');
    var filePathRev = path.join(root, fileNameRev);

    if(!fs.existsSync(filePathRev)){
        fs.writeFileSync(filePathRev, buffer);
        console.log('Copied ' + filePath + ' to ' + filePathRev);
    }
    return fileNameRev;
}

function identity(x) { return x; }


src/shader-manager.js
var Shader = require('./shader');

function ShaderManager(gl, sources, options){
    this.gl = gl;
    this.sources = sources;
    this.shaders = [];
    options = options || {};
    this.prefix = options.prefix || '';
    this.sourceIds = Object.create(null);
    this.sourceId = 1;
}
module.exports = ShaderManager;
ShaderManager.prototype = {
    includeExpression: /#include "([^"]+)"/,
    preprocess: function(name, content, included) {
        var sourceId = this.getSourceId(name);
        var lines = content.split('\n');
        var output = ['#line 1 ' + sourceId];
        for (var i = 0; i < lines.length; i++) {
            var line = lines[i];
            var match = line.match(this.includeExpression);
            if(match){
                var include = match[1];
                if(!included[include]){
                    output.push(this.getSource(include, {}, included));
                    output.push('#line ' + (i+2) + ' ' + sourceId);
                    included[include] = true;
                }
                else {
                    output.push('// ' + line);
                }
            }
            else {
                output.push(line);
            }
        }
        return output.join('\n');
    },
    getSource: function(name, defines, included) {
        name = name.replace(/\.\w+$/, '');
        var content = this.sources[this.prefix + name];
        if(content == null) {
            throw new Error('shader not found ' + name);
        }
        return glslDefines(defines) + this.preprocess(name, content, included);
    },
    getSourceId: function(name){
        if(!(name in this.sourceIds)){
            this.sourceIds[name] = this.sourceId++;
        }
        return this.sourceIds[name];
    },
    getSourceName: function(sourceId) {
        var sourceIds = this.sourceIds;
        for(var name in sourceIds){
            if(sourceIds[name] === sourceId) return name;
        }
    },
    get: function(vertex, frag, defines) {
        if(!frag) {
            frag = vertex;
        }
        var key = JSON.stringify([vertex, frag, defines]);
        if(!(key in this.shaders)){
            try {
                this.shaders[key] = new Shader(this.gl,
                    this.getSource(vertex, defines, {}),
                    this.getSource(frag, defines, {}));
            }
            catch(e){
                throw this.resolveError(e);
            }
        }
        return this.shaders[key];
    },
    resolveError: function(e) {
        if(!e.message || !e.message.match(/Shader (compiler|linker) error:/)) {
            return e;
        }
        var sourceIdExpression = /(ERROR: )(\d+)(:\d+:)/g;
        var message = e.message.replace(sourceIdExpression, function(_, head, sourceId, tail){
            var source = this.getSourceName(sourceId*1  ) || 'unknown-' + sourceId;
            return head + source + tail;
        }.bind(this));
        if(message === e.message) return e;
        return new Error(message);

    }
};

function glslDefines(defines){
    if(!defines) return '';
    var output = [];
    for(var key in defines){
        if(defines[key] != null){
            output.push('#define ' + key + ' ' + defines[key]);
        }
    }
    return output.join('\n');
}


demos/demo.js
/* globals $, normalmap */


// relight('gfx/logo.png', 'logo', {
//   brightness: 4.0,
//
// });
//
// relight('gfx/normals.png', 'c', {
//   brightness: 0.4,
//   metalness: 0.95,
//   roughness: 0.5,
//   ambient: 0.05
// });

// hacky hacky hack hack used to go single pass for some demos
window.isMobile =  /Mobile|android/i.test(navigator.userAgent);

function normalize(p){
    var scale = 1/Math.sqrt(p[0]*p[0]+p[1]*p[1]+p[2]*p[2]);
    p[0] *= -scale;
    p[1] *= -scale;
    p[2] *= -scale;
    return p;
}

function bindLight(lights, options){
    options = options || {};
    var left, top;
    var canvas = lights.canvas;
    var color = new Float32Array(options.color||[0.5, 0.5, 0.5]);
    var position = new Float32Array(3);

    var z = options.zOffset || 1;

    function resize(){
        var bounds = $(canvas).offset();
        top = bounds.top;
        left = bounds.left;
    }

    var lastX, lastY;
    lights.configure({onContextRestored: function(){
        render(lastX, lastY);
    }});

    function render(x, y){
        lastX = x;
        lastY = y;
        var p = position;
        if(options.directional) {
            p[0] = (x/window.innerWidth-0.5);
            p[1] = (y/window.innerHeight-0.5);
            p[2] = z;
            normalize(p);
        }
        else {
            p[0] = x/canvas.clientWidth;
            p[1] = y/canvas.clientHeight;
            p[2] = z;
        }
        if(options.directional) {

            lights.addDirectionalLight(p, color);
        }
        else {
            lights.addPointLight(p, color);
        }
        lights.render();
        // lights.addDirectionalLight(normalize(vec3(0.3, -1.3, 1.0)), vec3(0.1, 0.1, 0.1));
    }

    $(window).resize(resize);

    $(document)
        .on('mousemove', function(e){
            move(e.pageX, e.pageY);
        })
        .on('touchmove', function(e){
            move(e.originalEvent.touches[0].pageX, e.originalEvent.touches[0].pageY);
        });

    $(canvas).on('touchstart', function(e){ e.preventDefault(); });

    function move(x, y){
        if(options.directional) {
            render(x, y);
        }
        else {
            render(x - left, y - top);
        }
    }

    resize();
    render(canvas.width*0.25, canvas.height*0.1);
}

function logo(options) {
    $('#logo').normalmap({
        normalMap: 'gfx/logo/normal.png',
        materialMap: 'gfx/logo/material.png',
        ambient: 0,
        metalness: 1.0,
        roughness: 0.3,
        baseColor: normalmap.vec3([1.000, 0.766, 0.336])
    }).then(function(lights){
        bindLight(lights, $.extend({
            color: [0.8, 0.8, 0.8]
        }, options || {}));
    });
}
window.logo = logo;

if(window.location.hash == '#embed') {
    $('body').addClass('normalmap-demo-embedded');
    $('.normalmap-demo-nav a').each(function(){
        $(this).attr('href', $(this).attr('href') + '#embed');
    });
    $('.normalmap-demo-mobile-nav option').each(function(){
        $(this).attr('value', $(this).attr('value') + '#embed');
    });
}

if(location.protocol === 'file:'){
    window.alert('The normalmap.js demos need to be served from a webserver to work.');
}
